src/osiris_competing_read_coordinator.erl

```diff
@@ -0,0 +1,298 @@
+-define(DEFAULT_BATCH_SIZE, 3).
+
+-record(state, {log,
+                config,
+                sup,
+                readers,
+                service_queue,
+                reader_groups,
+                pending,
+                batch_size}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the server
+%% @end
+%%--------------------------------------------------------------------
+-spec start_link(term(), pid()) -> {ok, Pid :: pid()} |
+          {error, Error :: {already_started, pid()}} |
+          {error, Error :: term()} |
+          ignore.
+start_link(Config, SupPid) ->
+    gen_server:start_link(?MODULE, [Config, SupPid], []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Initializes the server
+%% @end
+%%--------------------------------------------------------------------
+-spec init(Args :: term()) -> {ok, State :: term()} |
+          {ok, State :: term(), Timeout :: timeout()} |
+          {ok, State :: term(), hibernate} |
+          {stop, Reason :: term()} |
+          ignore.
+init([Config, SupPid]) ->
+    process_flag(trap_exit, true),
+    gen_server:cast(self(), init),
+    {ok, #state{config = Config,
+                sup = SupPid,
+                readers = #{},
+                reader_groups = #{},
+                service_queue = queue:new(),
+                pending = [],
+                batch_size = application:get_env(osiris, crc_batch_size, ?DEFAULT_BATCH_SIZE)}}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling call messages
+%% @end
+%%--------------------------------------------------------------------
+-spec handle_call(Request :: term(), From :: {pid(), term()}, State :: term()) ->
+          {reply, Reply :: term(), NewState :: term()} |
+          {reply, Reply :: term(), NewState :: term(), Timeout :: timeout()} |
+          {reply, Reply :: term(), NewState :: term(), hibernate} |
+          {noreply, NewState :: term()} |
+          {noreply, NewState :: term(), Timeout :: timeout()} |
+          {noreply, NewState :: term(), hibernate} |
+          {stop, Reason :: term(), Reply :: term(), NewState :: term()} |
+          {stop, Reason :: term(), NewState :: term()}.
+handle_call({register, Pid, Tag, Prefetch, Fmt}, _From,
+            #state{readers = Readers,
+                   reader_groups = Groups0,
+                   service_queue = SQ} = State) ->
+    Reader = {Pid, Tag},
+    case maps:get(Reader, Readers, undefined) of
+        undefined ->
+            ReaderState = #{prefetch => Prefetch,
```

Comment:
I think I'd prefer a record for the reader state, it gives a little bit more help around tooling and that.